text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,315 @@
+    tokenizer: Box<dyn Tokenizer + 'static>,
+    normalizer: Box<dyn Normalizer + 'static>,
+}
+```
+
+#### Configuration
+
+The tokenizer is configured at initialization. A `Default` implementation is provided with the settings we believe suits most cases.
+The configuration options, for now, are:
+- the `stop_words`
+- the `pipeline_map` that links a `(Script, Language)` to a `Pipeline`. This defaults to the mapping we find best fitting for most cases. This settings should be left as is now, but will allow user defined Pipelines in the future.
+
+```rust
+pub struct AnalyzerConfig<'a, A> {
+    /// language specialized pipeline, this can be switched during
+    /// document tokenization if the document contains several languages
+    pub pipeline_map: HashMap<(Script, Language), Pipeline>,
+    pub stop_words: &'a Set<A>,
+}
+```
+
+#### Analyzer
+
+The analyzer exposes an abstracted interface to the tokenization process. Its API is standard:
+
+```rust
+use crate::token::Token;
+use crate::tokenizer::Tokenizer;
+
+pub struct Analyzer<'a, A> {
+    config: AnalyzerConfig<'a, A>,
+}
+
+impl<'a, A> Analyzer<'a, A>
+where
+    A: AsRef<[u8]>
+{
+    /// create a new tokenizer detecting script
+    /// and chose the specialized internal tokenizer
+    pub fn new(config: AnalyzerConfig<'a, A>) -> Self { unimplemented!() }
+
+    /// Builds an `AnalyzedText` instance with the correct analyzer pipeline, and pre-processes the
+    /// text.
+    ///
+    /// If an analysis pipeline exists for the inferred `(Script, Language)`, the analyzer will look
+    /// for a user specified default `(Script::Other, Language::Other)`. If the user default is not
+    /// specified, it will fallback to `(IdentityPreProcessor, UnicodeSegmenter, IdentityNormalizer)`.
+    ///
+    /// ```rust
+    /// use meilisearch_tokenizer::{Analyzer, AnalyzerConfig};
+    /// use fst::Set;
+    /// // defaults to unicode segmenter with identity preprocessor and normalizer.
+    /// let stop_words = Set::from_iter([""].iter()).unwrap();
+    /// let analyzer = Analyzer::new(AnalyzerConfig::default_with_stopwords(&stop_words));
+    /// let analyzed = analyzer.analyze("The quick (\"brown\") fox can't jump 32.3 feet, right? Brr, it's 29.3°F!");
+    /// let mut tokens = analyzed.tokens();
+    /// assert!("the" == tokens.next().unwrap().text());
+    /// ```
+    pub fn analyze<'t>(&'t self, text: &'t str) -> AnalyzedText<'t, A> { unimplemented!() }
+}
+
+/// result of analyzer.analyze() function, made to choose how tokens will be returned.
+pub struct AnalyzedText<'a, A> {}
+
+impl<'a, A> AnalyzedText<'a, A>
+where
+    A: AsRef<[u8]>
+{
+    /// Returns a `TokenStream` for the Analyzed text.
+    pub fn tokens(&'a self) -> TokenStream<'a> { unimplemented!() }
+
+    /// Attaches each token to its corresponding portion of the original text.
+    pub fn reconstruct(&'a self) -> impl Iterator<Item = (&'a str, Token<'a>)>  { unimplemented!() }
+}
+```
+
+#### Token
```

Comment:
Could you add a small explanation on the token? While I understand all the previous concepts this one I'm not sure. Is a token a portion of a string?